shesha-core/src/Shesha.Application/Notifications/NotificationSender.cs

```diff
@@ -0,0 +1,305 @@
+        /// <param name="toPerson"></param>
+        /// <param name="data"></param>
+        /// <param name="priority"></param>
+        /// <param name="attachments"></param>
+        /// <param name="triggeringEntity"></param>
+        /// <param name="channel"></param>
+        /// <returns></returns>
+        public async Task SendNotification<TData>(NotificationTypeConfig type, Person fromPerson, Person toPerson, TData data, RefListNotificationPriority priority, List<NotificationAttachmentDto> attachments = null, GenericEntityReference triggeringEntity = null, NotificationChannelConfig channel = null) where TData : NotificationData
+        {
+            var notification = await _notificationRepository.InsertAsync(new Notification()
+            {
+                NotificationType = type,
+                FromPerson = fromPerson,
+                ToPerson = toPerson,
+                NotificationData = JsonSerializer.Serialize(data),
+                TriggeringEntity = triggeringEntity,
+                Priority = priority
+            });
+
+            await _unitOfWorkManager.Current.SaveChangesAsync();
+
+            if (channel != null)
+            {
+                // Send notification to a specific channel
+                await SendNotificationToChannel(notification, data, fromPerson, toPerson, type, priority, channel, attachments);
+            }
+            else
+            {
+                // Send notification to all determined channels
+                var channels = await _notificationManager.GetChannelsAsync(type, toPerson, (RefListNotificationPriority)priority);
+
+                foreach (var channelConfig in channels)
+                {
+                    await SendNotificationToChannel(notification, data, fromPerson, toPerson, type, priority, channelConfig, attachments);
+                }
+            }
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <typeparam name="TData"></typeparam>
+        /// <param name="notification"></param>
+        /// <param name="data"></param>
+        /// <param name="fromPerson"></param>
+        /// <param name="toPerson"></param>
+        /// <param name="type"></param>
+        /// <param name="priority"></param>
+        /// <param name="channelConfig"></param>
+        /// <param name="attachments"></param>
+        /// <returns></returns>
+        /// <exception cref="UserFriendlyException"></exception>
+        private async Task SendNotificationToChannel<TData>(Notification notification, TData data, Person fromPerson, Person toPerson, NotificationTypeConfig type, RefListNotificationPriority priority, NotificationChannelConfig channelConfig, List<NotificationAttachmentDto> attachments = null) where TData : NotificationData
+        {
+            var template = await _messageTemplateRepository.FirstOrDefaultAsync(x => x.PartOf.Id == type.Id && channelConfig.SupportedFormat == x.MessageFormat);
```

Comment:
I think I mentioned it somewhere but can't find my comment. Domain model allows to store multiple template with the same message format. This code will work by a wrong way when there are more than one template with the same format. Options to solve it:
1. restrict creation of multiple templates with the same format
2. handle all existing templates
3. use single template but sort the query by some columns
Note: in any case the logic should be transparent, please review the requirements an choose one of ways